packetbeat/protos/dhcp/dhcp.go

```diff
@@ -0,0 +1,198 @@
+package dhcp
+
+import (
+	"bytes"
+	"encoding/hex"
+	"net"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/packetbeat/protos"
+)
+
+var hardwareTypes = map[int]string{
+	1:  "ethernet",
+	6:  "ieee_802",
+	7:  "arcnet",
+	11: "localtalk",
+	12: "localnet",
+	14: "smds",
+	15: "frame_relay",
+	16: "atm",
+	17: "hdlc",
+	18: "fibre_channel",
+	19: "atm",
+	20: "serial_line",
+}
+
+var messageTypes = []string{
+	"DHCPDISCOVER",
+	"DHCPOFFER",
+	"DHCPREQUEST",
+	"DHCPDECLINE",
+	"DHCPACK",
+	"DHCPNAK",
+	"DHCPRELEASE",
+	"DHCPINFORM",
+}
+
+type dhcpPlugin struct {
+	ports   []int
+	results protos.Reporter
+}
+
+func (dhcp *dhcpPlugin) init(results protos.Reporter, config *dhcpConfig) {
+	dhcp.ports = config.Ports
+	dhcp.results = results
+}
+
+func (dhcp *dhcpPlugin) GetPorts() []int {
+	return dhcp.ports
+}
+
+func payloadToRows(payload []byte) [][]byte {
+	result := make([][]byte, 0)
+	var idx = 0
+	for idx < len(payload) {
+		result = append(result, []byte{payload[idx], payload[idx+1], payload[idx+2], payload[idx+3]})
+		idx += 4
+	}
+	return result
+}
+
+func bytesToIPv4(bytes []byte) string {
+	ip := net.IPv4(bytes[0], bytes[1], bytes[2], bytes[3])
+	str := ip.String()
+	if str == "0.0.0.0" {
+		return ""
+	}
+	return str
+}
+
+func bytesToHardwareAddress(bytes []byte, length int) string {
+	return net.HardwareAddr(bytes[0:length]).String()
+}
+
+func combineRows(rows [][]byte, start int, length int) []byte {
+	combined := make([]byte, 0)
+	end := start + length
+	var idx = start
+	for idx < end {
+		combined = append(combined, rows[idx]...)
+		idx++
+	}
+	return combined
+}
+
+func trimNullBytes(b []byte) []byte {
+	index := bytes.Index(b, []byte{0})
+	return b[0:index]
+}
+
+func extractOptions(payload []byte) []byte {
+	result := make([]byte, 0)
+	var idx = 240
+	for idx < len(payload) {
+		result = append(result, payload[idx])
+		idx++
+	}
+	return result
+}
+
+func parseOptions(options []byte) map[int][]byte {
+	result := make(map[int][]byte)
+	var remaining = 0
+	var option = 0
+	var body = make([]byte, 0)
+	var idx = 0
+	for idx < len(options) {
+		if remaining == 0 {
+			if option != 0 {
+				result[option] = body
+				body = make([]byte, 0)
+			}
+			option = int(options[idx])
+			if option == 0 || option == 255 {
+				idx = len(options)
+			} else {
+				idx++
+				remaining = int(options[idx])
+			}
+		} else {
+			body = append(body, options[idx])
+			remaining--
```

Comment:
This function can be improved to read the option at once instead of looping and appending every byte:
```
remaining = int(options[idx])
if idx + remaining + 1 > len(options) {
   error...
}
result[option] = options[idx+1:idx+1+remaining]
```

As before, I'm not sure if it's necessary to copy into a new slice or not.

Reply:
I'm worried that this protocol can create a lot of noise on a busy network.

What do you think about adding a config option to only report dhcp messages based on their type. Something like:

report_types: [ 'DHCPACK', 'DHCPNACK' ]

So that only those two types get reported.

Also I'm not 100% sure which events would be interesting. Do ACK and NACK contain all the information (IP assignments) and do all transactions end in ACK or NACK?